File: ParaLensApp/libs/react-native-mlkit-ocr/android/src/main/java/com/margelo/nitro/mlkitocr/MlkitOcr.kt
package com.margelo.nitro.mlkitocr

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.YuvImage
import android.graphics.Rect
import android.util.Base64
import com.facebook.proguard.annotations.DoNotStrip
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.ByteBuffer
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

@DoNotStrip
class MlkitOcr : HybridMlkitOcrSpec() {
  private val textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
  
  override fun multiply(a: Double, b: Double): Double {
    return a * b
  }
  
  // MARK: - Static Image OCR Methods
  
  override fun recognizeText(imagePath: String): OcrResult {
    val latch = CountDownLatch(1)
    var ocrResult: OcrResult? = null
    var processingError: Exception? = null
    
    try {
      val file = File(imagePath)
      if (!file.exists()) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = "Image file not found: $imagePath"
        )
      }
      
      val bitmap = BitmapFactory.decodeFile(imagePath)
      if (bitmap == null) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = "Failed to decode image from path: $imagePath"
        )
      }
      
      val image = InputImage.fromBitmap(bitmap, 0)
      textRecognizer.process(image)
        .addOnSuccessListener { result ->
          ocrResult = processTextRecognitionResult(result)
          latch.countDown()
        }
        .addOnFailureListener { exception ->
          processingError = exception
          latch.countDown()
        }
      
      latch.await(10, TimeUnit.SECONDS)
      
      if (processingError != null) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = processingError!!.message ?: "Unknown error occurred"
        )
      }
      
      return ocrResult ?: OcrResult(
        text = "",
        blocks = emptyArray(),
        success = false,
        error = "Failed to process image"
      )
    } catch (e: Exception) {
      return OcrResult(
        text = "",
        blocks = emptyArray(),
        success = false,
        error = e.message ?: "Unknown error occurred"
      )
    }
  }
  
  override fun recognizeTextFromBase64(base64Image: String): OcrResult {
    val latch = CountDownLatch(1)
    var ocrResult: OcrResult? = null
    var processingError: Exception? = null
    
    try {
      val decodedBytes = Base64.decode(base64Image, Base64.DEFAULT)
      val bitmap = BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size)
      
      if (bitmap == null) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = "Failed to decode base64 image"
        )
      }
      
      val image = InputImage.fromBitmap(bitmap, 0)
      textRecognizer.process(image)
        .addOnSuccessListener { result ->
          ocrResult = processTextRecognitionResult(result)
          latch.countDown()
        }
        .addOnFailureListener { exception ->
          processingError = exception
          latch.countDown()
        }
      
      latch.await(10, TimeUnit.SECONDS)
      
      if (processingError != null) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = processingError!!.message ?: "Unknown error occurred"
        )
      }
      
      return ocrResult ?: OcrResult(
        text = "",
        blocks = emptyArray(),
        success = false,
        error = "Failed to process base64 image"
      )
    } catch (e: Exception) {
      return OcrResult(
        text = "",
        blocks = emptyArray(),
        success = false,
        error = e.message ?: "Unknown error occurred"
      )
    }
  }
  
  // MARK: - Vision Camera Frame Processing Methods
  
  override fun processFrame(frame: Frame): OcrResult {
    val latch = CountDownLatch(1)
    var ocrResult: OcrResult? = null
    var processingError: Exception? = null
    
    try {
      // Convert frame data to Bitmap
      val bitmap = convertFrameToBitmap(frame)
      if (bitmap == null) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = "Failed to convert frame to bitmap"
        )
      }
      
      val image = InputImage.fromBitmap(bitmap, 0)
      textRecognizer.process(image)
        .addOnSuccessListener { result ->
          ocrResult = processTextRecognitionResult(result)
          latch.countDown()
        }
        .addOnFailureListener { exception ->
          processingError = exception
          latch.countDown()
        }
      
      latch.await(10, TimeUnit.SECONDS)
      
      if (processingError != null) {
        return OcrResult(
          text = "",
          blocks = emptyArray(),
          success = false,
          error = processingError!!.message ?: "Unknown error occurred"
        )
      }
      
      return ocrResult ?: OcrResult(
        text = "",
        blocks = emptyArray(),
        success = false,
        error = "Failed to process frame"
      )
    } catch (e: Exception) {
      return OcrResult(
        text = "",
        blocks = emptyArray(),
        success = false,
        error = e.message ?: "Unknown error occurred"
      )
    }
  }
  
  override fun processFrameSync(frame: Frame): OcrResult {
    return processFrame(frame)
  }
  
  // MARK: - Utility Methods
  
  override fun isAvailable(): Boolean {
    return true
  }
  
  // MARK: - Private Helper Methods
  
  private fun convertFrameToBitmap(frame: Frame): Bitmap? {
    try {
      // Get the frame data as ArrayBuffer (actual runtime type may vary)
      val bufferObject = frame.toArrayBuffer()

      val bytes: ByteArray = when (bufferObject) {
        is ByteArray -> bufferObject
        is ByteBuffer -> {
          val duplicateBuffer = bufferObject.duplicate()
          duplicateBuffer.clear()
          val out = ByteArray(duplicateBuffer.remaining())
          duplicateBuffer.get(out)
          out
        }
        else -> return null
      }

      // Create Bitmap from byte array
      return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    } catch (e: Exception) {
      e.printStackTrace()
      return null
    }
  }
  
  private fun processTextRecognitionResult(result: com.google.mlkit.vision.text.Text): OcrResult {
    val allText = StringBuilder()
    val blocks = mutableListOf<TextBlock>()
    
    for (block in result.textBlocks) {
      val blockText = block.text
      allText.append(blockText).append("\n")
      
      val boundingBox = block.boundingBox
      val textBlock = TextBlock(
        text = blockText,
        boundingBox = BoundingBox(
          left = boundingBox?.left?.toDouble() ?: 0.0,
          top = boundingBox?.top?.toDouble() ?: 0.0,
          right = boundingBox?.right?.toDouble() ?: 0.0,
          bottom = boundingBox?.bottom?.toDouble() ?: 0.0
        ),
        cornerPoints = if (boundingBox != null) {
          arrayOf(
            CornerPoint(x = boundingBox.left.toDouble(), y = boundingBox.top.toDouble()),
            CornerPoint(x = boundingBox.right.toDouble(), y = boundingBox.top.toDouble()),
            CornerPoint(x = boundingBox.right.toDouble(), y = boundingBox.bottom.toDouble()),
            CornerPoint(x = boundingBox.left.toDouble(), y = boundingBox.bottom.toDouble())
          )
        } else {
          emptyArray()
        }
      )
      blocks.add(textBlock)
    }
    
    return OcrResult(
      text = allText.toString().trim(),
      blocks = blocks.toTypedArray(),
      success = true,
      error = null
    )
  }
}

File: ParaLensApp/libs/react-native-mlkit-ocr/ios/MlkitOcr.swift
import Foundation
import MLKitTextRecognition
import MLKitVision
import UIKit

class MlkitOcr: HybridMlkitOcrSpec {
    private let textRecognizer = TextRecognizer.textRecognizer()

    public func multiply(a: Double, b: Double) throws -> Double {
        return a * b
    }

    // MARK: - Static Image OCR Methods

    public func recognizeText(imagePath: String) throws -> OcrResult {
        let semaphore = DispatchSemaphore(value: 0)
        var ocrResult: OcrResult?
        var processingError: Error?

        Task {
            do {
                guard let image = UIImage(contentsOfFile: imagePath) else {
                    ocrResult = OcrResult(
                        text: "",
                        blocks: [],
                        success: false,
                        error: "Failed to load image from path: \(imagePath)"
                    )
                    semaphore.signal()
                    return
                }

                let visionImage = VisionImage(image: image)
                let result = try await textRecognizer.process(visionImage)

                ocrResult = processTextRecognitionResult(result)
            } catch {
                processingError = error
            }
            semaphore.signal()
        }

        semaphore.wait()

        if let error = processingError {
            throw error
        }

        return ocrResult ?? OcrResult(
            text: "",
            blocks: [],
            success: false,
            error: "Failed to process image"
        )
    }

    public func recognizeTextFromBase64(base64Image: String) throws -> OcrResult {
        let semaphore = DispatchSemaphore(value: 0)
        var ocrResult: OcrResult?
        var processingError: Error?

        Task {
            do {
                guard let imageData = Data(base64Encoded: base64Image),
                      let image = UIImage(data: imageData) else {
                    ocrResult = OcrResult(
                        text: "",
                        blocks: [],
                        success: false,
                        error: "Failed to decode base64 image"
                    )
                    semaphore.signal()
                    return
                }

                let visionImage = VisionImage(image: image)
                let result = try await textRecognizer.process(visionImage)

                ocrResult = processTextRecognitionResult(result)
            } catch {
                processingError = error
            }
            semaphore.signal()
        }

        semaphore.wait()

        if let error = processingError {
            throw error
        }

        return ocrResult ?? OcrResult(
            text: "",
            blocks: [],
            success: false,
            error: "Failed to process base64 image"
        )
    }

    // MARK: - Vision Camera Frame Processing Methods

    public func processFrame(frame: Frame) throws -> OcrResult {
        let semaphore = DispatchSemaphore(value: 0)
        var ocrResult: OcrResult?
        var processingError: Error?

        Task {
            do {
                // Convert frame data to UIImage
                let image = try convertFrameToUIImage(frame)
                let visionImage = VisionImage(image: image)
                let result = try await textRecognizer.process(visionImage)

                ocrResult = processTextRecognitionResult(result)
            } catch {
                processingError = error
            }
            semaphore.signal()
        }

        semaphore.wait()

        if let error = processingError {
            throw error
        }

        return ocrResult ?? OcrResult(
            text: "",
            blocks: [],
            success: false,
            error: "Failed to process frame"
        )
    }

    public func processFrameSync(frame: Frame) throws -> OcrResult {
        return try processFrame(frame)
    }

    // MARK: - Utility Methods

    public func isAvailable() throws -> Bool {
        return true
    }

    // MARK: - Private Helper Methods

    private func convertFrameToUIImage(_ frame: Frame) throws -> UIImage {
        // Get the frame data as ArrayBuffer
        let arrayBuffer = frame.toArrayBuffer()
        
        // Convert ArrayBuffer to Data
        let data = Data(bytes: arrayBuffer, count: arrayBuffer.count)
        
        // Create UIImage from data
        guard let image = UIImage(data: data) else {
            throw NSError(domain: "MlkitOcr", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to create UIImage from frame data"])
        }
        
        return image
    }

    private func processTextRecognitionResult(_ result: Text) -> OcrResult {
        var allText = ""
        var blocks: [TextBlock] = []

        for block in result.blocks {
            let blockText = block.text
            allText += blockText + "\n"

            let textBlock = TextBlock(
                text: blockText,
                boundingBox: BoundingBox(
                    left: block.frame.origin.x,
                    top: block.frame.origin.y,
                    right: block.frame.origin.x + block.frame.size.width,
                    bottom: block.frame.origin.y + block.frame.size.height
                ),
                cornerPoints: [
                    CornerPoint(x: block.frame.origin.x, y: block.frame.origin.y),
                    CornerPoint(x: block.frame.origin.x + block.frame.size.width, y: block.frame.origin.y),
                    CornerPoint(x: block.frame.origin.x + block.frame.size.width, y: block.frame.origin.y + block.frame.size.height),
                    CornerPoint(x: block.frame.origin.x, y: block.frame.origin.y + block.frame.size.height)
                ]
            )
            blocks.append(textBlock)
        }

        return OcrResult(
            text: allText.trimmingCharacters(in: .whitespacesAndNewlines),
            blocks: blocks,
            success: true
        )
    }
}
